### PR TITLE
Improve lambda options and introduce deprecation

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**
Add additional options to control formatting of C++11 lambdas (see changes for details).

The new options are named in such a way that the existing `sp_cpp_lambda_paren` becomes unfortunately named. Consequently, I have taken the opportunity to finally start to do something about #1815, as I have been wanting to do for ages, and implemented a mechanism for deprecating options. Because #1815 is a journey rather than a destination, this doesn't exactly "fix" it, but it does show how we can accomplish #1815's goals going forward. (At least as far as a simple rename. I hope to have patches for more complicated deprecations as well at some point.)

### CODE REVIEW

**Comment by guy-maurel** [comment]
@CDanU , @mihaipopescu : 
Please comment the PR as you have an important part at the #1815

File: src/option.cpp [hunk 1]
```diff
@@ -898,9 +940,29 @@ void process_option_line(const std::string &config_line, const char *filename)
          }
       }
    }
+   else if (cmd == "using")
+   {
+      auto vargs = split_args(args[1], filename, is_varg_sep);
+      if (vargs.size() != 2)
+      {
+         OptionWarning w{ filename };
+         w("%s requires a version number in the form MAJOR.MINOR", cmd.c_str());
```
Comment (CDanU) [inline src/option.cpp]: Previously we had to release uncrustify-0.66.1 and uncrustify-0.68.1.
Why don't we use PATCH here as well?

File: src/option.h [hunk 2]
```diff
@@ -144,8 +144,14 @@ class GenericOption
 class OptionWarning
 {
 public:
-   OptionWarning(const char *filename);
-   OptionWarning(const GenericOption *);
+   enum Severity
```
Comment (CDanU) [inline src/option.h]: Do you have a reason for not using enum classes here?

File: src/option.cpp [hunk 3]
```diff
@@ -898,9 +940,29 @@ void process_option_line(const std::string &config_line, const char *filename)
          }
       }
    }
+   else if (cmd == "using")
```
Comment (CDanU) [inline src/option.cpp]: I am not entirely sure about the `using` keyword here. I see a small possibility for confusion as there are already concepts bound to that keyword in languages that we are formatting.

**Comment by CDanU** [comment]
The more I think about it, the more I think that we don't actually need a directive at all. 

Provided that we don't flip flop on the changes ( no `a -> b -> a` changes, only a -> b -> c) we should be able to deduce the version (or at least roughly a state) based on the options available in the config.

**Comment by mwoehlke-kitware** [comment]
> The more I think about it, the more I think that we don't actually need a directive at all. [...] we should be able to deduce the version (or at least roughly a state) based on the options available in the config.

Since configs can be sparse, there is no guarantee that an option that would signify the version is present. This means we wouldn't be able to e.g. change the meaning of an option, which I believe is something we want to be able to do.

**Comment by mwoehlke-kitware** [comment]
> I wish a documentation file with explanations how to use the new feature.

I can probably do that. I take it you mean something like a "how do I deprecate an option" guide? (Seems like a good candidate for a wiki page?)

Thanks!

**Pull request merged**